sdk/python/kfp/_client.py

```diff
@@ -178,29 +178,34 @@ def _extract_pipeline_yaml(self, tar_file):
       with tar.extractfile(all_yaml_files[0]) as f:
         return yaml.load(f)
 
-  def run_pipeline(self, experiment_id, job_name, pipeline_package_path, params={}):
+  def run_pipeline(self, experiment_id, job_name, pipeline_package_path=None, params={}, pipeline_id=None):
     """Run a specified pipeline.
 
     Args:
       experiment_id: The string id of an experiment.
       job_name: name of the job.
       pipeline_package_path: local path of the pipeline package(tar.gz file).
       params: a dictionary with key (string) as param name and value (string) as as param value.
+      pipeline_id: the string ID of a pipeline.
 
     Returns:
       A run object. Most important field is id.
     """
     import kfp_run
 
-    pipeline_obj = self._extract_pipeline_yaml(pipeline_package_path)
-    pipeline_json_string = json.dumps(pipeline_obj)
+    pipeline_json_string = None
+    if pipeline_package_path:
+      pipeline_obj = self._extract_pipeline_yaml(pipeline_package_path)
+      pipeline_json_string = json.dumps(pipeline_obj)
     api_params = [kfp_run.ApiParameter(name=_k8s_helper.K8sHelper.sanitize_k8s_name(k), value=str(v))
                   for k,v in params.items()]
     key = kfp_run.models.ApiResourceKey(id=experiment_id,
                                         type=kfp_run.models.ApiResourceType.EXPERIMENT)
     reference = kfp_run.models.ApiResourceReference(key, kfp_run.models.ApiRelationship.OWNER)
     spec = kfp_run.models.ApiPipelineSpec(
-        workflow_manifest=pipeline_json_string, parameters=api_params)
+        pipeline_id=pipeline_id,
```

Comment:
when both the pipeline_id and workflow_manifest are provided, does the backend choose the pipeline_id over the other?

Reply:
I am not sure. I think it should fail with 400 error. Client should let backend to decide whether to throw or not.

Reply:
It will fail. Then the backend error will be propagated to the sdk. Fair enough.